src/redis.c

```diff
@@ -67,43 +100,82 @@ void
 redis_producer_produce(Producer p, Message msg)
 {
     Meta m = (Meta)p->meta;
-    m->reply = redisCommand(m->c, "LPUSH %s %s",m->topic, (char *) message_get_data(msg));
-    freeReplyObject(m->reply);
+
+    if (m->pipe_max == 0) { /* No pipelining. */
+        m->reply = redisCommand(m->c, "LPUSH %s %s",m->topic, (char *) message_get_data(msg));
+        freeReplyObject(m->reply);
+    } else { /* Pipelining */
+        redisAppendCommand(m->c, "LPUSH %s %s",m->topic, (char *) message_get_data(msg));
+        m->pipe_cur++;
+        redis_meta_check_pipeline(m, true);
+    }
 }
 
 void
 redis_producer_free(Producer *p)
 {
     Meta m = (Meta) ((*p)->meta);
+    redis_meta_check_pipeline(m, false);
     redis_meta_free(&m);
     free(*p);
     *p = NULL;
 }
 
 Consumer
-redis_consumer_init(char *host, char *topic)
+redis_consumer_init(char *host, char *topic, int pipeline)
 {
     Consumer redis = calloc(1, sizeof(*redis));
 
-    redis->meta          = redis_meta_init(host, topic);
+    redis->meta          = redis_meta_init(host, topic, pipeline);
     redis->consumer_free = redis_consumer_free;
     redis->consume       = redis_consumer_consume;
 
     return redis;
 }
 
+static void
+redis_consumer_handle_reply(const redisReply *reply, Message msg)
+{
+    if (reply->type == REDIS_REPLY_ARRAY && reply->elements == 2)
+    {
+        char *result = calloc(reply->element[1]->len + 1, sizeof(*result));
+        strncpy(result, reply->element[1]->str, reply->element[1]->len);
+        message_set_data(msg, result);
+    }
+}
+
 int
 redis_consumer_consume(Consumer c, Message msg)
 {
     Meta m = (Meta)c->meta;
-    m->reply = redisCommand(m->c, "BLPOP %s 1", m->topic);
-    if (m->reply->type == REDIS_REPLY_ARRAY && m->reply->elements == 2)
-    {
-        char *result = calloc(m->reply->element[1]->len + 1, sizeof(*result));
-        strncpy(result, m->reply->element[1]->str, m->reply->element[1]->len);
-        message_set_data(msg, result);
+
+    if (m->pipe_max == 0) {
+        /* No pipelining. */
+        m->reply = redisCommand(m->c, "BLPOP %s 1", m->topic);
+        redis_consumer_handle_reply(m->reply, msg);
         freeReplyObject(m->reply);
+        return 0;
     }
+
+    /* Pipelining */
+    if (m->pipe_cur < m->pipe_max) {
+        redisAppendCommand(m->c, "BLPOP %s 1", m->topic);
+        m->pipe_cur++;
+    }
+
+    if (m->pipe_cur == m->pipe_max)
```

Comment:
I would put a >= there instead just for future bug proofing.

Reply:
Good point. fixup changes that to `>=`